mm2src/kdf_walletconnect/src/client/stream.rs

```diff
@@ -0,0 +1,293 @@
+
+    /// Error generated when failed to parse an inbound message, invalid request
+    /// type or message ID.
+    InboundError(ClientError),
+
+    /// Error generated when failed to write data to the underlying websocket
+    /// stream.
+    OutboundError(ClientError),
+
+    /// The websocket connection was closed.
+    ///
+    /// This is the last event that can be produced by the stream.
+    ConnectionClosed(Option<CloseFrame<'static>>),
+}
+
+/// Lower-level [`FusedStream`] interface for the client connection.
+///
+/// The stream produces [`StreamEvent`] when polled, and can be used to send RPC
+/// requests (see [`ClientStream::send()`] and [`ClientStream::send_raw()`]).
+///
+/// For a higher-level interface see [`Client`](crate::client::Client). For an
+/// example usage of the stream see `client::connection` module.
+pub struct ClientStream {
+    socket: WebSocketStream,
+    socket_ended: bool,
+    outbound_tx: UnboundedSender<Message>,
+    outbound_rx: UnboundedReceiver<Message>,
+    requests: HashMap<MessageId, oneshot::Sender<Result<serde_json::Value, ClientError>>>,
+    id_generator: MessageIdGenerator,
+    close_frame: Option<CloseFrame<'static>>,
+}
+
+/// Opens a connection to the Relay and returns [`ClientStream`] for the
+/// connection.
+pub async fn open_new_relay_connection_stream(
+    request: &HttpRequest<()>,
+) -> MmResult<ClientStream, WebsocketClientError> {
+    let stream = connect(request.uri().to_string())
+        .await
+        .map_to_mm(|err| WebsocketClientError::TransportError(err.to_string()))?;
+    Ok(ClientStream::new(stream))
+}
+
+impl ClientStream {
+    fn new(socket: WebSocketStream) -> Self {
+        let id_generator = MessageIdGenerator::new();
+        let (outbound_tx, outbound_rx) = unbounded_channel();
+        let requests = HashMap::new();
+
+        Self {
+            socket,
+            outbound_rx,
+            outbound_tx,
+            requests,
+            close_frame: None,
+            id_generator,
+            socket_ended: false,
+        }
+    }
+
+    /// Sends an already serialized [`OutboundRequest`][OutboundRequest] (see
+    /// [`create_request()`]).
+    pub fn send_raw(&mut self, request: OutboundRequest) {
+        let tx = request.tx;
+        let id = self.id_generator.next();
+        let request = Payload::Request(rpc::Request::new(id, request.params));
+        let serialized = serde_json::to_string(&request);
+
+        match serialized {
+            Ok(data) => match self.requests.entry(id) {
+                Entry::Occupied(_) => {
+                    tx.send(Err(ClientError::DuplicateRequestId)).ok();
+                },
+
+                Entry::Vacant(entry) => {
+                    entry.insert(tx);
+                    self.outbound_tx.send(Message::Text(data)).ok();
+                },
+            },
+
+            Err(err) => {
+                tx.send(Err(ClientError::SerdeError(err.to_string()))).ok();
```

Comment:
i think we better call this `deserialization error` instead of `serde`.

Reply:
I used SerdeError to cover both `Des` and `Ser`.. 🤔

Reply:
done